### ISSUE TITLE
Runtime numbers are unreliable

### ISSUE THREAD

**Issue by athas** [gradbench/gradbench#142]
GradBench currently evaluates each function only once on each workload.  This makes the runtime measurements highly sensitive to noise, and also unrealistically pessimistic, as caches will be cold. This becomes even more significant once you want to measure performance on GPUs and the like. Accurately benchmarking computer programs is a complicated problem, but I don't think GradBench needs to go to a lot of effort to get numbers that are reasonably reliable. E.g. ADBench just does ten runs per workload and takes the average, and this is *fine*.

But how should this be implemented? Should the eval do ten (or whatever) evaluate messages? Should the intermediary automatically duplicate these messages? Should the tools internally do multiple runs and report the average? Or multiple runs and report all of the runtimes, to enable statistical analysis?

I have a mild preference for letting the tool do multiple runs internally, because it cuts down on the potentially costly communication of large datasets. It also makes it easier to "keep the cache warm" for tools where that matters (e.g., GPUs, where you might normally deallocate all data after evaluation). If we provide a bit of reusable Python and C++ glue code, then it is not so difficult to accomplish this in a consistent way across all tools. I would suggest that the tool then reports the runtimes of all the evaluations, in order to maximise the information available for further analysis.

**Comment by samestep** [comment]
The functions in the eval can be defined however the eval wants. Since we're trusting the tools to be a bit honest, can't we just change the evals to add an additional integer parameter in the `"input"` of each function, specifying how many times that function should be evaluated by the tool before returning the outputs from just one of those evaluations?

**Comment by samestep** [comment]
I think another issue here is that we don't have control over what runner GitHub uses for each individual job in the `run` matrix, so it seems quite possible that different kinds of runners will happen to be used for different tools. There are a couple ways I can think of to try to address this:

1. Replace the `matrix` in the `run` job with just a single job, that iterates over all the eval/tool pairs and runs them all in sequence. This would be slower, but at least we guarantee they all run in the same VM.
2. Use [self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners) to guarantee the exact machine configuration for every `run` job. This is more expensive and complex, but allows us to maintain consistency across workflow runs rather than just within a single workflow run. I think @avibryant mentioned something about this in the past.

**Comment by athas** [comment]
> The functions in the eval can be defined however the eval wants. Since we're trusting the tools to be a bit honest, can't we just change the evals to add an additional integer parameter in the "input" of each function, specifying how many times that function should be evaluated by the tool before returning the outputs from just one of those evaluations? Then the eval could make this parameter configurable via https://github.com/gradbench/gradbench/pull/127.

Yes, this would be a good way to do it, as it still puts the multiple-runs logic inside each tool, thus enabling cache warmup etc. It is fine to just return one output for validation (say, the last one), but I think it is best to return multiple runtimes.

> I think another issue here is that we don't have control over what runner GitHub uses for each individual job in the run matrix, so it seems quite possible that different kinds of runners will happen to be used for different tools.

Any runtime numbers coming out of GA will always be suspect since they are run on shared virtual machines. It's a very noisy environment. It's fine to put the numbers on the website as a rough measurement, but any *real* results, reported in papers or whatnot, should be done in a controlled environment. To me, the main purpose of the GA setup is to ensure that things *work*, not to produce fully reliable numbers.

I have experience using self-hosted runners, and they work very well. The main problem is actually having that hardware somewhere. I suggest dealing with that once GradBench is a wealthy and powerful project.

**Comment by athas** [comment]
I also have a machine in a cabinet here that can operate as a self-hosted runner. They really aren't very difficult to set up.

If someone has a spare M1 Mac Mini that you can put on a shelf at CMU, then that could also solve the cross compilation woes.

**Comment by samestep** [comment]
Yeah; I guess currently I'm leaning toward first making it easy for people to run all the benchmarks locally on their own machine, unless we have a good reason to work on self-hosted runners at the moment. I'm sure you're right that they're not too bad to set up, but it's the ongoing maintenance that I'm a bit wary of.

Regarding an M1 Mac: just one wouldn't really help since we already get 5 concurrent macOS jobs for free via GitHub Actions, as I wrote in https://github.com/gradbench/gradbench/pull/153#issuecomment-2532655870.

**Comment by athas** [comment]
Does GA run macOS jobs on real ARM hardware? Then why do we cross compile at all?

**Comment by samestep** [comment]
Yep! And the reason we cross-compile is that I don't know how to build separate parts of a cross-platform Docker image on different machines and then combine them. So if I understand correctly, unless there's a way to do that, we need to cross-compile using either QEMU or Rosetta, and currently we just happen to be using QEMU because GitHub has a higher concurrency limit for Linux jobs.

**Comment by athas** [comment]
The DiffSharp failures are due to qemu limitations, based on my brief investigation.

**Comment by samestep** [comment]
@athas this seems resolved to me. Should we close this issue?

**Comment by athas** [comment]
Yeah, this is solved as well as it can be. The main related problem is that CI uses too few workloads for most evals.
